File: frontend/src/api/finance.ts
import client from './client'
import type { ApiResponse } from '../types/common'

export type FinancePeriod = 'daily' | 'weekly' | 'monthly'

export type FinanceSummary = {
  period: FinancePeriod
  total_revenue_rmb: number | null
  total_cost_rmb: number | null
  net_profit_rmb: number | null
  profit_margin_pct: number | null
  cash_balance_rmb: number | null
  entry_count: number
  cost_breakdown: Record<string, number>
  platform_settlement: PlatformSettlementSummary
  risk_signals?: FinanceRiskSignal[]
  data_status: 'ready' | 'data_required'
  source_refs?: Array<Record<string, unknown>>
  evidence_window?: string
  confidence_reason?: string
  data_gaps?: string[]
}

export type FinanceRiskSignal = {
  code: string
  level: 'info' | 'medium' | 'high' | string
  title: string
  detail: string
  action_label: string
  action_route: string
}

export type PlatformSettlementSummary = {
  wallet_balances: PlatformWalletBalance[]
  movement_totals: Record<string, number>
  order_reconciliation: {
    linked_order_count: number
    linked_entry_count: number
  }
}

export type PlatformWalletBalance = {
  platform: string | null
  market: string | null
  amount_rmb: number
  amount_original: number | null
  currency: string
  account_name: string | null
  reference_rate: string | null
  source_entry_id: string
  occurred_at: string | null
}

export type FinanceTraceback = {
  period: FinancePeriod
  summary: {
    order_count: number
    product_count: number
    store_count: number
    entry_count: number
    total_revenue_rmb: number | null
    total_cost_rmb: number | null
    net_profit_rmb: number | null
    refund_rmb: number
    platform_bill_rmb: number
    settlement_movement_rmb: number
  }
  by_order: FinanceTracebackOrder[]
  by_product: FinanceTracebackProduct[]
  by_store: FinanceTracebackStore[]
  data_status: 'ready' | 'data_required'
}

export type FinanceTracebackBase = {
  platform: string | null
  market: string | null
  revenue_rmb: number | null
  cost_rmb: number | null
  net_profit_rmb: number | null
  cost_breakdown: Record<string, number>
  entry_count: number
  source_entry_ids: string[]
  data_gaps: string[]
}

export type FinanceTracebackOrder = FinanceTracebackBase & {
  order_id: string
  account_name: string | null
}

export type FinanceTracebackProduct = FinanceTracebackBase & {
  product_id: string
  product_name: string
  v5_sku_contexts?: Array<{
    status?: 'matched' | 'unmatched' | string
    source?: 'v5_product_sku_variants' | string
    platform_listing_id?: string | null
    sku_variant_id?: string
    ledger_sku?: string | null
    merchant_sku?: string
    platform_sku?: string | null
    spu?: string | null
    skc?: string | null
    option_1?: { name?: string | null; value?: string | null } | null
    option_2?: { name?: string | null; value?: string | null } | null
    listing_stock?: number | null
    listing_price?: number | null
    source_entry_id?: string
    available_sku_count?: number
    data_gaps?: string[]
  }>
}

export type FinanceTracebackStore = FinanceTracebackBase & {
  store_key: string
  account_name: string | null
}

export type FinanceLedgerCreate = {
  entry_type: string
  amount_rmb: number
  amount_original?: number | null
  currency?: string
  platform?: string | null
  market?: string | null
  order_id?: string | null
  sourcing_item_id?: string | null
  description?: string | null
  extra?: Record<string, unknown>
  occurred_at?: string | null
}

export type FinanceLedgerEntry = FinanceLedgerCreate & {
  id: string
  currency: string
  extra: Record<string, unknown>
  occurred_at: string
}

export type PlatformBillImportRecord = {
  import_ref?: string | null
  entry_type: string
  amount_rmb: number
  amount_original?: number | null
  currency?: string
  platform?: string | null
  market?: string | null
  order_id?: string | null
  sourcing_item_id?: string | null
  account_name?: string | null
  product_name?: string | null
  description?: string | null
  occurred_at?: string | null
}

export type PlatformBillImportResult = {
  imported_count: number
  skipped_count: number
  imported_entry_ids: string[]
  skipped: Array<{ import_ref: string | null; reason: string }>
}

export type PlatformBillSyncResult = {
  sync_log_id: string
  status: string
  platform_account_id: string
  platform: string
  account_name: string
  connection_status?: string | null
  implementation_status?: string | null
  import_result: PlatformBillImportResult
  data_gaps: string[]
  message?: string | null
  next_action?: string | null
}

export type FinanceEntryTypeOption = {
  id: string
  label: string
  source: 'dictionary' | 'history'
}

export async function getFinanceSummary(period: FinancePeriod, params?: { platform_account_id?: string }) {
  const res = await client.get<ApiResponse<FinanceSummary>>('/finance/summary', { params: { period, ...params } })
  return res.data
}

export async function getFinanceTraceback(period: FinancePeriod, params?: { platform_account_id?: string }) {
  const res = await client.get<ApiResponse<FinanceTraceback>>('/finance/traceback', { params: { period, ...params } })
  return res.data
}

export async function listFinanceLedger(params?: { page?: number; page_size?: number; entry_type?: string; platform_account_id?: string; order_id?: string }) {
  const res = await client.get<ApiResponse<FinanceLedgerEntry[]>>('/finance/ledger', { params })
  return res.data
}

export async function listFinanceEntryTypes() {
  const res = await client.get<ApiResponse<FinanceEntryTypeOption[]>>('/finance/entry-types')
  return res.data
}

export async function createFinanceLedger(data: FinanceLedgerCreate) {
  const res = await client.post<ApiResponse<any>>('/finance/ledger', data)
  return res.data
}

export async function importPlatformBills(data: { records: PlatformBillImportRecord[] }) {
  const res = await client.post<ApiResponse<PlatformBillImportResult>>('/finance/platform-bills/import', data)
  return res.data
}

export async function syncPlatformBills(data: { platform_account_id: string; start_at?: string | null; end_at?: string | null }) {
  const res = await client.post<ApiResponse<PlatformBillSyncResult>>('/finance/platform-bills/sync', data)
  return res.data
}

File: backend/app/schemas/finance.py
"""Schemas for finance ledger APIs."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class FinanceLedgerCreate(BaseModel):
    entry_type: str = Field(..., min_length=1, max_length=50)
    amount_rmb: float
    amount_original: Optional[float] = None
    currency: str = "CNY"
    platform: Optional[str] = None
    market: Optional[str] = None
    order_id: Optional[str] = None
    sourcing_item_id: Optional[str] = None
    description: Optional[str] = None
    extra: dict = Field(default_factory=dict)
    occurred_at: Optional[datetime] = None


class FinanceLedgerResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    entry_type: str
    amount_rmb: float
    amount_original: Optional[float] = None
    currency: str
    platform: Optional[str] = None
    market: Optional[str] = None
    order_id: Optional[str] = None
    sourcing_item_id: Optional[str] = None
    description: Optional[str] = None
    extra: dict
    occurred_at: datetime


class PlatformBillImportRecord(BaseModel):
    import_ref: Optional[str] = None
    entry_type: str = Field(..., min_length=1, max_length=50)
    amount_rmb: float
    amount_original: Optional[float] = None
    currency: str = "CNY"
    platform: Optional[str] = None
    market: Optional[str] = None
    order_id: Optional[str] = None
    sourcing_item_id: Optional[str] = None
    account_name: Optional[str] = None
    product_name: Optional[str] = None
    description: Optional[str] = None
    occurred_at: Optional[datetime] = None


class PlatformBillImportRequest(BaseModel):
    records: list[PlatformBillImportRecord] = Field(..., min_length=1, max_length=200)


class PlatformBillImportSkipped(BaseModel):
    import_ref: Optional[str] = None
    reason: str


class PlatformBillImportResponse(BaseModel):
    imported_count: int
    skipped_count: int
    imported_entry_ids: list[str] = Field(default_factory=list)
    skipped: list[PlatformBillImportSkipped] = Field(default_factory=list)


class PlatformBillSyncRequest(BaseModel):
    platform_account_id: str = Field(..., min_length=1)
    start_at: Optional[datetime] = None
    end_at: Optional[datetime] = None


class PlatformBillSyncResponse(BaseModel):
    sync_log_id: str
    status: str
    platform_account_id: str
    platform: str
    account_name: str
    connection_status: Optional[str] = None
    implementation_status: Optional[str] = None
    import_result: PlatformBillImportResponse
    data_gaps: list[str] = Field(default_factory=list)
    message: Optional[str] = None
    next_action: Optional[str] = None


class PlatformWalletBalance(BaseModel):
    platform: Optional[str] = None
    market: Optional[str] = None
    amount_rmb: float
    amount_original: Optional[float] = None
    currency: str
    account_name: Optional[str] = None
    reference_rate: Optional[str] = None
    source_entry_id: str
    occurred_at: Optional[str] = None


class OrderReconciliationSummary(BaseModel):
    linked_order_count: int
    linked_entry_count: int


class PlatformSettlementSummary(BaseModel):
    wallet_balances: list[PlatformWalletBalance] = Field(default_factory=list)
    movement_totals: dict[str, float] = Field(default_factory=dict)
    order_reconciliation: OrderReconciliationSummary


class FinanceRiskSignal(BaseModel):
    code: str
    level: str
    title: str
    detail: str
    action_label: str
    action_route: str


class FinanceTracebackSummary(BaseModel):
    order_count: int
    product_count: int
    store_count: int
    entry_count: int


class FinanceOrderTraceback(BaseModel):
    order_id: str
    platform: Optional[str] = None
    market: Optional[str] = None
    account_name: Optional[str] = None
    revenue_rmb: Optional[float] = None
    cost_rmb: Optional[float] = None
    net_profit_rmb: Optional[float] = None
    cost_breakdown: dict[str, float] = Field(default_factory=dict)
    entry_count: int
    source_entry_ids: list[str] = Field(default_factory=list)
    data_gaps: list[str] = Field(default_factory=list)


class FinanceProductTraceback(BaseModel):
    product_id: str
    product_name: str
    platform: Optional[str] = None
    market: Optional[str] = None
    revenue_rmb: Optional[float] = None
    cost_rmb: Optional[float] = None
    net_profit_rmb: Optional[float] = None
    cost_breakdown: dict[str, float] = Field(default_factory=dict)
    entry_count: int
    source_entry_ids: list[str] = Field(default_factory=list)
    data_gaps: list[str] = Field(default_factory=list)


class FinanceStoreTraceback(BaseModel):
    store_key: str
    platform: Optional[str] = None
    market: Optional[str] = None
    account_name: Optional[str] = None
    revenue_rmb: Optional[float] = None
    cost_rmb: Optional[float] = None
    net_profit_rmb: Optional[float] = None
    cost_breakdown: dict[str, float] = Field(default_factory=dict)
    entry_count: int
    source_entry_ids: list[str] = Field(default_factory=list)
    data_gaps: list[str] = Field(default_factory=list)


class FinanceTracebackResponse(BaseModel):
    period: str
    summary: FinanceTracebackSummary
    by_order: list[FinanceOrderTraceback] = Field(default_factory=list)
    by_product: list[FinanceProductTraceback] = Field(default_factory=list)
    by_store: list[FinanceStoreTraceback] = Field(default_factory=list)
    data_status: str


class FinanceSummaryResponse(BaseModel):
    period: str
    total_revenue_rmb: Optional[float]
    total_cost_rmb: Optional[float]
    net_profit_rmb: Optional[float]
    profit_margin_pct: Optional[float]
    cash_balance_rmb: Optional[float]
    entry_count: int
    cost_breakdown: dict[str, float]
    platform_settlement: PlatformSettlementSummary
    risk_signals: list[FinanceRiskSignal] = Field(default_factory=list)
    data_status: str
